src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,243 @@
+  } else if constexpr (numIndexColumns == 2) {
+    return std::tie(lt->triple_.ids_[1], lt->triple_.ids_[2]);
+  } else {
+    return std::tie(lt->triple_.ids_[2]);
+  }
+}
+
+// ____________________________________________________________________________
+template <size_t numIndexColumns>
+IdTable LocatedTriplesPerBlock::mergeTriplesImpl(size_t blockIndex,
+                                                 const IdTable& block) const {
+  // This method should only be called if there are located triples in the
+  // specified block.
+  AD_CONTRACT_CHECK(map_.contains(blockIndex));
+
+  AD_CONTRACT_CHECK(numIndexColumns <= block.numColumns());
+  AD_CONTRACT_CHECK(block.numColumns() >= 1);
+
+  auto numInsertsAndDeletes = numTriples(blockIndex);
+  IdTable result{block.numColumns(), block.getAllocator()};
+  result.resize(block.numRows() + numInsertsAndDeletes.numAdded_);
+
+  const auto& locatedTriples = map_.at(blockIndex);
+
+  auto cmpLt = [](auto lt, auto row) {
+    return tieIdTableRow<numIndexColumns>(row) >
+           tieLocatedTriple<numIndexColumns>(lt);
+  };
+  auto cmpEq = [](auto lt, auto row) {
+    return tieIdTableRow<numIndexColumns>(row) ==
+           tieLocatedTriple<numIndexColumns>(lt);
+  };
+
+  auto row = block.begin();
+  auto locatedTriple = locatedTriples.begin();
+  auto resultEntry = result.begin();
+
+  auto writeTripleToResult = [&result, &resultEntry](auto& locatedTriple) {
+    for (size_t i = 0; i < numIndexColumns; i++) {
+      (*resultEntry)[i] = locatedTriple.triple_.ids_[3 - numIndexColumns + i];
+    }
+    // Write UNDEF to any additional columns.
+    for (size_t i = numIndexColumns; i < result.numColumns(); i++) {
+      (*resultEntry)[i] = ValueId::makeUndefined();
+    }
+    resultEntry++;
+  };
+
+  while (row != block.end() && locatedTriple != locatedTriples.end()) {
+    if (cmpLt(locatedTriple, *row)) {
+      // locateTriple < row
+      if (locatedTriple->shouldTripleExist_) {
+        // Insertion of a non-existent triple.
+        writeTripleToResult(*locatedTriple);
+      }
+      locatedTriple++;
+    } else if (cmpEq(locatedTriple, *row)) {
+      // locateTriple == row
+      if (!locatedTriple->shouldTripleExist_) {
+        // Deletion of an existing triple.
+        row++;
+      }
+      locatedTriple++;
+    } else {
+      // locateTriple > row
+      // The row is not deleted - copy it
+      *resultEntry++ = *row++;
+    }
+  }
+
+  if (locatedTriple != locatedTriples.end()) {
+    AD_CORRECTNESS_CHECK(row == block.end());
+    std::ranges::for_each(
+        std::ranges::subrange(locatedTriple, locatedTriples.end()) |
+            std::views::filter(&LocatedTriple::shouldTripleExist_),
+        writeTripleToResult);
+  }
+  if (row != block.end()) {
+    AD_CORRECTNESS_CHECK(locatedTriple == locatedTriples.end());
+    while (row != block.end()) {
+      *resultEntry++ = *row++;
+    }
+  }
+
+  result.resize(resultEntry - result.begin());
+  return result;
+}
+
```

Comment:
Add our usual `// ______________________________________________________________` comments above all implementations.

Reply:
(Also for many other functions in this file)